lib/functoria/engine.ml

```diff
@@ -81,28 +83,24 @@ let files info t stage =
    the functor [c] applies to [args]. *)
 let module_expression fmt (c, args) =
   Fmt.pf fmt "%s%a" (Device.module_name c)
-    Fmt.(list (parens @@ of_to_string @@ Device_graph.impl_name))
+    Fmt.(list (parens @@ of_to_string @@ Device.Graph.impl_name))
     args
 
 let find_all_devices info g i =
   let ctx = Info.context info in
   let id = Impl.with_left_most_device ctx i { f = Device.id } in
-  let p = function
-    | Device_graph.Dev d -> Device.id d = id
-    | App | If _ -> false
+  let f x l =
+    let (Device.Graph.D { dev; _ }) = x in
+    if Device.id dev = id then x :: l else l
```

Comment:
@Drup are you sure we really want to select all devices with the same id here? I don't remember why we need to need to find "all devices" here. The `connect` code mainly try to associate the init devices by their counterpart in the graph.

Reply:
`find_all_devices` was added rather recently with the `~init` parameter. I'm personally not super convinced we need to code it like that, but I wasn't sure, so I preserved the same semantics as was present before.

Since we are very careful never to duplicate any devices, I think this list always ever have one element ...

Reply:
but what happens if you add the same device twice in the graph?

Reply:
That function should probably just be a call to `with_left_most_device` ...